core/servicedata/service.go

```diff
@@ -0,0 +1,111 @@
+package servicedata
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/goto/shield/core/project"
+	"github.com/goto/shield/core/relation"
+	"github.com/goto/shield/core/resource"
+	"github.com/goto/shield/core/user"
+	"github.com/goto/shield/internal/schema"
+	"github.com/goto/shield/pkg/uuid"
+)
+
+const keyNamespace = "shield/servicedata_key"
+
+type ResourceService interface {
+	Create(ctx context.Context, res resource.Resource) (resource.Resource, error)
+}
+
+type RelationService interface {
+	Create(ctx context.Context, rel relation.RelationV2) (relation.RelationV2, error)
+}
+
+type ProjectService interface {
+	Get(ctx context.Context, idOrSlug string) (project.Project, error)
+}
+
+type UserService interface {
+	FetchCurrentUser(ctx context.Context) (user.User, error)
+}
+
+type Service struct {
+	repository      Repository
+	resourceService ResourceService
+	relationService RelationService
+	projectService  ProjectService
+	userService     UserService
+}
+
+func NewService(repository Repository, resourceService ResourceService, relationService RelationService, projectService ProjectService, userService UserService) *Service {
+	return &Service{
+		repository:      repository,
+		resourceService: resourceService,
+		relationService: relationService,
+		projectService:  projectService,
+		userService:     userService,
+	}
+}
+
+func (s Service) CreateKey(ctx context.Context, key Key) (Key, error) {
+	// check if key contains ':'
+	if strings.Contains(key.Key, ":") || key.Key == "" {
+		return Key{}, ErrInvalidDetail
+	}
+
+	// fetch current user
+	currentUser, err := s.userService.FetchCurrentUser(ctx)
+	if err != nil {
+		return Key{}, fmt.Errorf("%w: %s", user.ErrInvalidEmail, err.Error())
+	}
+
+	// convert project slug to project id
+	if !uuid.IsValid(key.ProjectID) {
+		project, err := s.projectService.Get(ctx, key.ProjectID)
+		if err != nil {
+			return Key{}, err
+		}
+		key.ProjectID = project.ID
+	}
+
+	// create URN
+	key.URN = key.CreateURN()
+
+	// insert the service data key
+	resource, err := s.resourceService.Create(ctx, resource.Resource{
+		Name:        key.URN,
+		NamespaceID: keyNamespace,
+		ProjectID:   key.ProjectID,
+		UserID:      currentUser.ID,
+	})
+	if err != nil {
+		return Key{}, err
+	}
+	key.ResourceID = resource.Idxa
+
+	// insert service data key to the servicedata_keys table
+	createdServiceDataKey, err := s.repository.CreateKey(ctx, key)
```

Comment:
This should use transaction. In case create relation error, this should roll back.

Reply:
we can create a separate PR for this